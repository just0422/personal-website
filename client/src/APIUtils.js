import axios from 'axios';

const fileHeader = { headers: { 'Content-Type': 'multipart/form-data' } }

const apiPath = '/api/v1';
export const projectsPath = '/projects';
export const experiencesPath = '/experiences';
export const skillsPath = '/skills';
export const commentsPath = '/comments';
export const screenshotsPath = '/screenshots';
export const contactPath = '/contact'

export const projectsUrl = apiPath + projectsPath;
export const experiencesUrl = apiPath + experiencesPath;
export const skillsUrl = apiPath + skillsPath;
export const contactUrl = apiPath + contactPath;
export const resetUrl = '/reset';

export default {
  projects() {
    return {
      getAll: () => axios.get(projectsUrl),
      getSkills: id => axios.get(`${projectsUrl}/${id}${skillsPath}`),
      getComments: id => axios.get(`${projectsUrl}/${id}${commentsPath}`),
			getScreenshot: (project_id, screenshot_id) => axios.get(`${projectsUrl}/${project_id}${screenshotsPath}/${screenshot_id}`),
      getScreenshots: id => axios.get(`${projectsUrl}/${id}${screenshotsPath}`),
      createScreenshot: (id, screenshot) => axios.post(`${projectsUrl}/${id}${screenshotsPath}`, screenshot, fileHeader),
      updateProject: (id, project) => axios.patch(`${projectsUrl}/${id}`, project),
			deleteComment: (project_id, comment_id) => axios.delete(`${projectsUrl}/${project_id}${commentsPath}/${comment_id}`),
    };
  },

  experiences() {
    return {
      getAll: () => axios.get(experiencesUrl),
      getSkills: id => axios.get(`${experiencesUrl}/${id}${skillsPath}`),
      getComments: id => axios.get(`${experiencesUrl}/${id}${commentsPath}`),
    };
  },

  skills() {
    return {
      getAll: () => axios.get(skillsUrl),
    };
	},

	contact() {
		return {
			emailContact: (contact) => axios.post(contactUrl, contact),
		}
	},

	reset() {
		return {
			get: () => axios.get(resetUrl)
		}
	}
};
