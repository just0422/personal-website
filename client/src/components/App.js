import React, {Component} from 'react';
import {Button} from 'react-bootstrap';
import {PacmanLoader} from 'react-spinners';
import Countdown from 'react-countdown-now';

import NavigationBar from './NavigationBar';
import Main from './Main';
import ErrorModal from 'Error';

import api from 'APIUtils';
import '../stylesheets/App.scss';

class App extends Component {
  constructor(props) {
    super(props);

    this.handleLoading = this.handleLoading.bind(this);
    this.resetTime = this.resetTime.bind(this);
    this.resetMessage = this.resetMessage.bind(this);

    this.state = {
      countDownTo: -1,
      loading: false,
      resetText: 'Resetting in ',
      timeToReset: false
    };
  }

  handleLoading() {
    this.setState({loading: true});

    api
      .reset()
      .get()
      .then(resp => {
        this.setState({loading: false});
      })
      .catch(err => {
        this.setState({
          loading: false,
          error: err,
        });
      });
  }

  resetMessage() {
    this.setState({
      timeToReset: true,
      resetText: 'Website data is resetting. Please refresh soon!'
    })
  }

  resetTime() {
    const interval = 900000; // 15 min
    let time = Date.now();
    let remainder = time % interval;

    this.setState({
      countDownTo: time + interval - remainder,
    });
  }

  componentDidMount() {
    this.resetTime();
  }

  render() {
    if (this.state.error) {
      return <ErrorModal component="Reset" error={this.state.error} />;
    } else if (this.state.loading) {
      return (
        <div>
          <h1>Website is resetting...</h1>
          <h3>Please hold. This takes about 15 seconds.</h3>

          <PacmanLoader
            sizeUnit={'px'}
            size={150}
            color={'#a00'}
            loading={this.state.loading}
          />
        </div>
      );
    } else {
      return (
        <div className="App">
          <NavigationBar />
          <Main />
          <div className="website-reset-countdown">
            {!this.state.timeToReset &&
              <Button id="reset-button" onClick={this.handleLoading}>
                Reset Now
              </Button>
            }
            {' ' + this.state.resetText}
            {!this.state.timeToReset &&
              <Countdown
                date={this.state.countDownTo}
                onComplete={this.resetMessage}
                precision={2}
                zeroPadDays={0}
              />
            }
          </div>
        </div>
      );
    }
  }
}

export default App;
