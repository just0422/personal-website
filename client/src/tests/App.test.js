import React, {Component} from 'react';
import {shallow, mount} from 'enzyme';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import App from 'App';
import ErrorModal from 'Error';

import {resetUrl} from 'APIUtils';

describe('App test', () => {
  let mock;

  const flushPromises = () => new Promise(resolve => setTimeout(resolve));
  const resetFail = () => mock.onGet(resetUrl).reply(500);
  const resetSucceed = () => mock.onGet(resetUrl).reply(200);

  beforeEach(() => (mock = new MockAdapter(axios)));

  it('should render correctly', () => {
    const component = shallow(<App />);

    expect(component).toMatchSnapshot();
    expect(component.state('countDownTo')).toBeGreaterThan(-1);
  });

  it('should handle reset click', () => {
    const component = shallow(<App />);

    let reset = component.find('Button#reset-button');
    reset.simulate('click');

    expect(component).toMatchSnapshot();
    expect(component.state('loading')).toBe(true);
  });

  it('should handle reset submit and succeed', async () => {
    resetSucceed();
    const component = shallow(<App />);

    let reset = component.find('Button#reset-button');
    reset.simulate('click');
    await flushPromises();
    component.update();

    expect(component).toMatchSnapshot();
    expect(component.state('loading')).toBe(false);
  });

  it('should handle reset submit and fail', async () => {
    resetFail();
    const component = shallow(<App />);

    let reset = component.find('Button#reset-button');
    reset.simulate('click');
    await flushPromises();
    component.update();

    expect(component).toMatchSnapshot();
    expect(component.state('loading')).toBe(false);
    expect(component.state('error')).toBeTruthy();
    expect(component.containsMatchingElement(<ErrorModal />)).toBe(true);
  });
});
