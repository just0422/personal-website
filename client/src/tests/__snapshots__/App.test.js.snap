// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test should handle reset click 1`] = `ShallowWrapper {}`;

exports[`App test should handle reset submit and fail 1`] = `ShallowWrapper {}`;

exports[`App test should handle reset submit and succeed 1`] = `ShallowWrapper {}`;

exports[`App test should render correctly 1`] = `ShallowWrapper {}`;
